test/s3-mock.ts

```diff
@@ -0,0 +1,60 @@
+import { Readable } from 'stream';
+import type { S3 } from '@aws-sdk/client-s3';
+import { afterEach } from '@jest/globals';
+import type { S3Url } from '../lib/modules/datasource/maven/s3';
+
+let mockedObjects: Record<string, string> = {};
+let mockedTimestamps: Record<string, string> = {};
+
+function objectKey(url: S3Url) {
+  return `s3://${url.Bucket}/${url.Key}`;
+}
+
+function headObject(url: S3Url) {
+  const k = objectKey(url);
+  if (mockedObjects[k] === undefined) {
+    return Promise.reject({ message: 'NotFound' });
+  }
+  const LastModified = mockedTimestamps[k];
+  return Promise.resolve({ LastModified });
+}
+
+function getObject(url: S3Url) {
+  const k = objectKey(url);
+  if (!mockedObjects[k]) {
+    return Promise.reject({ message: 'NotFound' });
+  }
+  const Body = new Readable();
+  const content = mockedObjects[k];
+  if (typeof content === 'string') {
+    Body.push(content);
+  }
+  Body.push(null);
+  const LastModified = mockedTimestamps[k];
+  return Promise.resolve({ Body, LastModified });
+}
+
+function mockObject(url: string, content?: string, headers?: any) {
+  mockedObjects[url] = content ? content : '';
+  if (headers?.['Last-Modified']) {
+    mockedTimestamps[url] = headers['Last-Modified'];
+  }
+}
+
+const s3mock = {
+  headObject,
+  getObject,
+  mockObject,
+};
+export { s3mock };
+
+jest.mock('@aws-sdk/client-s3', () => ({
```

Comment:
I think `jest.mock` usage is problematic here, as it's acting kind of per-module.
Most likely, this very call causes further relying on module load order and resorting to `eslint-disable-next-line import/order` for `maven/index.spec.ts`.

I would rather recommend to extract `() => ({ S3: ... })` and export it as named `s3mockFactory`, so that it would be possible to use in `index.spec.ts`, e.g.:
```ts
jest.mock('@aws-sdk/client-s3', s3mockFactory)
```

Reply:
@zharinov I gave this a go in [this branch](https://github.com/kennethjor/renovate/compare/kennethjor/s3-support...kennethjor/s3-support-mock-refactor), but it doesn't work.

```
    expect(received).toEqual(expected) // deep equality

    Expected: {"display": "org.example:package", "group": "org.example", "name": "package", "registryUrl": "s3://repobucket", "releases": [{"version": "0.0.1"}, {"version": "1.0.0"}, {"version": "1.0.1"}, {"version": "1.0.2"}, {"version": "1.0.3-SNAPSHOT"}, {"version": "2.0.0"}]}
    Received: null

      288 |     const res = await get('org.example:package', baseUrlS3);
      289 |
    > 290 |     expect(res).toEqual({
          |                 ^
      291 |       display: 'org.example:package',
      292 |       group: 'org.example',
      293 |       name: 'package',

      at Object.<anonymous> (lib/modules/datasource/maven/index.spec.ts:290:17)

```

Reply:
I think `aws-sdk-client-mock` should be used, as it makes mocking aws sdk more easy

Reply:
@zharinov @viceice Couldn't we perhaps merge the mocking as-is? It does work and isn't completely silly. If someone would like to refactor it later, someone who understands Jest and friends better than I, they can do that.

I've been working on this PR for quite a while now and would very much like to have Maven dependencies from S3 managed by Renovate.

Reply:
nope, this code doesn't look good to me, sorry.
maybe I can find some time to help with the mock lib.